### PR TITLE
Jacks remorse - add "discard changes" confirmation when closing forms with unsaved data

### PR DESCRIPTION
**Pull request by robguthrie**
Needs good smoke testing before deploy.

Also all dropdowns need to be checked because we updated angular-bootstrap-ui

### CODE REVIEW

File: lineman/bower.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "faye": "latest",
     "angular-translate": "~2.7.2",
     "angular-translate-loader-url": "~2.7.2",
-    "angular-bootstrap": "0.11.2",
+    "angular-bootstrap": "0.13.2",
```
Comment (robguthrie) [inline lineman/bower.json]: this upgrade means we need to check all the dropdowns.. some of them have a buggy structure

File: lineman/app/models/discussion_model.coffee [hunk 2]
```diff
@@ -90,6 +102,9 @@ angular.module('loomioApp').factory 'DiscussionModel', (BaseModel) ->
     isStarred: ->
       @reader().starred
 
+    isUnsaved: ->
```
Comment (gdpelican) [inline lineman/app/models/discussion_model.coffee]: Always true? ?

File: lineman/app/services/form_service.coffee [hunk 3]
```diff
@@ -1,6 +1,14 @@
-angular.module('loomioApp').factory 'FormService', ($rootScope, FlashService) ->
+angular.module('loomioApp').factory 'FormService', ($rootScope, FlashService, $filter) ->
   new class FormService
 
+    confirmDiscardChanges: (event, record) ->
+      if record.isModified()
+        console.log record.modifiedAttributes()
+        if !confirm($filter('translate')('common.confirm_discard_changes'))
```
Comment (gdpelican) [inline lineman/app/services/form_service.coffee]: I was thinking it felt a little clunky to have entirely new modals for things like deactivate group, deactivate user, remove coordinator, etc. I wonder if we could use this pattern in those places too? Would probably feel better than spinning up a whole new modal here, and I'm thinking the behaviour should be the same for this thing and those things.

File: lineman/app/components/thread_page/thread_page.haml [hunk 4]
```diff
@@ -18,8 +18,8 @@
         .thread-notification-level.lmo-btn-wrapper
           %notification_volume_dropdown{translate-root: 'discussion', discussion: 'threadPage.discussion'}
         %star_toggle{thread: 'threadPage.discussion'}
-        %span.lmo-btn-wrapper.dropdown{ng-if: 'threadPage.showContextMenu()'}
-          %button.btn.lmo-btn-nude.dropdown-toggle{href:''}
+        %span.lmo-btn-wrapper{dropdown: '',  ng-if: 'threadPage.showContextMenu()'}
```
Comment (gdpelican) [inline lineman/app/components/thread_page/thread_page.haml]: Hmm, does this mean we need to go fix other dropdowns?

File: lineman/app/components/discussion_form/discussion_form.coffee [hunk 5]
```diff
@@ -1,27 +1,25 @@
...
+
     $scope.discussion = discussion.clone()
 
     actionName = if $scope.discussion.isNew() then 'created' else 'updated'
+
     $scope.submit = FormService.submit $scope, $scope.discussion,
       flashSuccess: "discussion_form.messages.#{actionName}"
-      successCallback: (response) => $location.path "/d/#{response.discussions[0].key}" if actionName == 'created'
+      successCallback: (response) =>
+        $location.path "/d/#{response.discussions[0].key}" if actionName == 'created'
```
Comment (gdpelican) [inline lineman/app/components/discussion_form/discussion_form.coffee]: LmoUrlService here?

Comment (gdpelican) [inline lineman/app/components/discussion_form/discussion_form.coffee]: (Oh this is just a line movement change, I see)

File: lineman/app/components/discussion_form/discussion_form.coffee [hunk 6]
```diff
@@ -1,27 +1,25 @@
-angular.module('loomioApp').factory 'DiscussionForm', ->
+angular.module('loomioApp').factory 'DiscussionForm', (Records) ->
```
Comment (gdpelican) [inline lineman/app/components/discussion_form/discussion_form.coffee]: Is this a stray Records injection?

File: lineman/app/components/discussion_form/discussion_form.coffee [hunk 7]
```diff
@@ -1,27 +1,25 @@
...
-      $scope.discussion.group().discussionPrivacyOptions if $scope.discussion.group()
-
-    $scope.setCurrentPrivacy = ->
-      if $scope.discussion.isNew()
-        $scope.discussion.private = $scope.getCurrentPrivacy() == 'private_only'
-
-    $scope.setCurrentPrivacy()
-
     $scope.showPrivacyForm = ->
-      $scope.getCurrentPrivacy() == 'public_or_private'
```
Comment (gdpelican) [inline lineman/app/components/discussion_form/discussion_form.coffee]: For some reason I get really obsessive about boolean functions, so ignore me if you don't care :+1: 

```
$scope.discussion.groupId and $scope.discussion.group().privacy == 'public_or_private'
```

**Comment by gdpelican** [comment]
The best! :bear:

File: lineman/bower.json (hunk 1 above)
Comment (gdpelican) [inline lineman/bower.json]: Ah ok. I think quite a few of em are worked out at least once with the integration tests, yay!

File: lineman/app/components/discussion_form/discussion_form.coffee (hunk 6 above)
Comment (robguthrie) [inline lineman/app/components/discussion_form/discussion_form.coffee]: yes! Thanks for that.

File: lineman/app/components/discussion_form/discussion_form.coffee (hunk 5 above)
Comment (robguthrie) [inline lineman/app/components/discussion_form/discussion_form.coffee]: Do you think we should use it here?

File: lineman/app/components/discussion_form/discussion_form.coffee (hunk 7 above)
Comment (robguthrie) [inline lineman/app/components/discussion_form/discussion_form.coffee]: thanks. I'm down with that.. some of this is misshapen after I tried putting console.log every step of the way

File: lineman/app/components/thread_page/thread_page.haml (hunk 4 above)
Comment (robguthrie) [inline lineman/app/components/thread_page/thread_page.haml]: Yes indeed. I've recorded that in the context just now.

**Comment by robguthrie** [comment]
Thanks for the quick review! Super helpful.

**Comment by robguthrie** [comment]
the following TODO's still exist for this branch:

* write e2e tests
* enable unsaved check on comment form
* fix proposal closing at always being modified by form
* Smoke testing discussion, prop, vote, comment new/edit

However the recordStore upgrade branch was based on this one... (the isModified stuff got me hooked) and it's getting really close to passing all e2es.. so I think I'll add those todos to that branch and close this.

**Pull request closed without merging**